### PR TITLE
lib/db, lib/syncthing: Don't repair DB on upgrade, but on error (fixes #6917)

### LINKED ISSUE #6917: Reconsider db check on upgrade
It used to be quite quick and painless; seconds on my computer. Now up to about five minutes on my computer. Other setups are worse:

```
<sphalerite> Is there any way to make "Checking db due to upgrade" faster? It takes hours on my phone.
```

I think it's the block loading that now happens...

### PR DESCRIPTION
**Pull request by imsodin**
No longer repair the db unconditionally on upgrade (fixes #6917). Instead when an error due to an inconsistency occurs, mark the db as needing repair. Then on next startup, do the repair.

### CODE REVIEW

**Comment by calmh** [comment]
I think this is reasonable, but it doesn't feel like something that needs to leak outside the db package. I suggest:

- setting the repair bit outside the database itself, because it may not be writable or the write+panic might make it not flush (maybe?). Something like touching `databaseDir+".needsrepair"`?

- having the lowlevel itself check for this on open and run the repair before returning the database

**Comment by calmh** [comment]
Oh and perhaps we should add a repair cycle before running migrations, but that's halfway unrelated.

File: lib/db/lowlevel.go [hunk 1]
```diff
@@ -82,6 +86,14 @@ func NewLowlevel(backend backend.Backend, opts ...Option) *Lowlevel {
 	}
 	db.keyer = newDefaultKeyer(db.folderIdx, db.deviceIdx)
 	db.Add(util.AsService(db.gcRunner, "db.Lowlevel/gcRunner"))
+	if loc := backend.Location(); loc != "" {
+		path := filepath.Join(loc, needsRepairFile)
```
Comment (calmh) [inline lib/db/lowlevel.go:90]: We should probably not assume that `loc` is a directory that we can do stuff in. We should be able to tack on an extension to loc though

File: lib/db/lowlevel.go [hunk 2]
```diff
@@ -82,6 +86,14 @@ func NewLowlevel(backend backend.Backend, opts ...Option) *Lowlevel {
...
+		if _, err := os.Lstat(path); err == nil {
+			l.Infoln("Database was marked for repair - this may take a while")
+			db.CheckRepair()
```
Comment (calmh) [inline lib/db/lowlevel.go:93]: CheckRepair no longer needs to be exported I think

Comment (imsodin) [reply lib/db/lowlevel.go]: A model test was using it, but luckily that was replaceable by `WithRecheckInterval(time.Millisecond)`.

**Review by calmh — APPROVED** [review]

**Pull request merged**